File: week5/Homework.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


# 参考文献

# 線分の交差判定
# https://qiita.com/wihan23/items/03efd7cd40dfec96a987

# 焼きなましについて
# https://shindannin.hatenadiary.com/entry/2021/03/06/115415


# In[2]:


#!/usr/bin/env python3

import sys
import math
import random
import time
import csv
import copy


# In[3]:


# calculate a probability for simulated annealing
def getProbability(start_time, time_limit, new_total_distance, total_distance, initial_distance):
    if (new_total_distance<total_distance):
        return 1.0
    start_temp = initial_distance*0.025;
    end_temp = initial_distance*0.0075;
    temp = start_temp + (end_temp - start_temp) * (time.time()/(start_time+time_limit))
    probability = math.exp((total_distance-new_total_distance)/temp)
    return probability


# In[4]:


# calculate a distance between the two points
def distance(city1, city2):
    return math.sqrt((city1[0] - city2[0]) ** 2 + (city1[1] - city2[1]) ** 2)


# In[5]:


# swap two points inside the route
def swapRoute(cities, tour, total_distance, first, second):
    
    new_tour = tour[:]
    new_total_distance = total_distance
    
    subtract1 = distance (cities[new_tour[first]], cities[new_tour[first+1]])
    subtract2 = distance (cities[new_tour[second]], cities[new_tour[second+1]])
    subtract3 = distance (cities[new_tour[first]], cities[new_tour[first-1]])
    subtract4 = distance (cities[new_tour[second]], cities[new_tour[second-1]])
    
    temp = new_tour[first]
    new_tour[first] = new_tour[second]
    new_tour[second] = temp
    
    add1 = distance (cities[new_tour[first]], cities[new_tour[first+1]])
    add2 = distance (cities[new_tour[second]], cities[new_tour[second+1]])
    add3 = distance (cities[new_tour[first]], cities[new_tour[first-1]])
    add4 = distance (cities[new_tour[second]], cities[new_tour[second-1]])
    
    new_total_distance -= (subtract1 + subtract2 + subtract3 + subtract4)
    new_total_distance += (add1 + add2 + add3 + add4)
    
    return new_tour, new_total_distance


# In[6]:


# checking whether x or y coordinates of two lines (created by p1&p2 and p3&p4) has an intersection
def max_min_cross(p1, p2, p3, p4):
    min_ab = min(p1, p2)
    max_ab = max(p1, p2)
    min_cd = min(p3, p4)
    max_cd = max(p3, p4)

    if min_ab > max_cd or max_ab < min_cd:
        return False

    return True


# checking whether two lines (created by p1&p2 and p3&p4 has an intersection)
def judgeCross(i, j, cities, tour):
    first = cities[tour[i]]
    second = cities[tour[j]]
    first_next = cities[tour[i+1]]
    second_next = cities[tour[j+1]]
    # x座標による判定
    if not max_min_cross(first[0], first_next[0], second[0], second_next[0]):
        return False

    # y座標による判定
    if not max_min_cross(first[1], first_next[1], second[1], second_next[1]):
        return False

    tc1 = (first[0] - first_next[0]) * (second[1] - first[1]) + (first[1] - first_next[1]) * (first[0] - second[0])
    tc2 = (first[0] - first_next[0]) * (second_next[1] - first[1]) + (first[1] - first_next[1]) * (first[0] - second_next[0])
    td1 = (second[0] - second_next[0]) * (first[1] - second[1]) + (second[1] - second_next[1]) * (second[0] - first[0])
    td2 = (second[0] - second_next[0]) * (first_next[1] - second[1]) + (second[1] - second_next[1]) * (second[0] - first_next[0])
    return tc1 * tc2 <= 0 and td1 * td2 <= 0


# In[7]:


# when the two lines intersect each other, reorganize the points so that there is no intersections
def noCross(cities, tour, total_distance):
    continue_flag = True
    while continue_flag:
        continue_flag = False
        N = len(cities)
        for i in range (1, N-2):
            for j in range(i+2, N-1):
                if judgeCross(i, j, cities, tour):
                    #from IPython.core.debugger import Pdb; Pdb().set_trace()
                    tour, temp_distance = fixChain(cities, tour, total_distance, i, j)
                    if temp_distance<total_distance:
                        continue_flag = True
                        total_distance = temp_distance
    
    return tour, total_distance


# In[8]:


# changing the order of the tour and re-calculate the distance of the given two points
def fixChain(cities, tour, total_distance, start, end):
    new_tour = tour[:]
    new_total_distance = 0
    for i in range (end-start):
        new_tour[start+i+1]=tour[end-i]
        
    N = len(cities)
    
    for i in range (N-1):
        new_total_distance += distance(cities[new_tour[i]], cities[new_tour[i+1]])
        
    return new_tour, new_total_distance


# In[9]:


# simulated annealing
def annealing(cities, tour, total_distance, initial_distance, start_time, time_limit):
    first = random.randrange(len(tour))
    second = random.randrange(len(tour))
    
    if first==0 or second ==0:
        return tour, total_distance
    if first==len(tour)-1 or second==len(tour)-1:
        return tour, total_distance
    if first==second:
        return tour, total_distance
    
    new_tour, new_total_distance = swapRoute(cities, tour, total_distance, first, second) #TODO
    
    rand = random.random()
    probability = getProbability(start_time, time_limit, new_total_distance, total_distance, initial_distance)
    
    print(str(probability)+" , "+str(rand)+" , "+str(new_total_distance)+" , "+str(total_distance))
    
#     Annealing1
    if probability>rand:
        tour = new_tour
        total_distance = new_total_distance

#   Annealing2
#     if new_total_distance < total_distance:
#         tour = new_tour
#         total_distance = new_total_distance
    
    return tour, total_distance


# In[10]:


# creating an initial answer based on greedy technique
def greedy(cities):
    N = len(cities)

    dist = [[0] * N for i in range(N)]
    for i in range(N):
        for j in range(i, N):
            dist[i][j] = dist[j][i] = distance(cities[i], cities[j])

    current_city = 0
    unvisited_cities = set(range(1, N))
    tour = [current_city]
    total_distance = 0
    
    while unvisited_cities:
        next_city = min(unvisited_cities,
                        key=lambda city: dist[current_city][city])
        total_distance += dist[current_city][next_city]
        unvisited_cities.remove(next_city)
        tour.append(next_city)
        current_city = next_city
        
        
    total_distance += dist[current_city][tour[0]]
    tour.append(tour[0])
    return tour, total_distance


# In[11]:


# do brute-force method if the number of cities is small
def brute(cities, tour, total_distance):
    if len(cities)<9:
        now_place = 0
        visited = [False] * len(cities)
        visited[0]=True
        count_visited = 1
        now_distance = 0
        now_tour = [0]
        tour, total_distance = run_brute(cities, tour, now_tour, now_distance, now_place, visited, count_visited, total_distance)
    
    return tour, total_distance


def run_brute (cities, tour, now_tour, now_distance, now_place, visited, count_visited, min_distance):
    if count_visited == len(cities):
        now_distance += distance(cities[now_place], cities[0])
        now_tour.append(0)
        if now_distance<min_distance:
            tour = now_tour
            min_distance = now_distance
        return tour, min_distance
        
    for i in range (len(cities)):
        if visited[i]:
            continue
            
        visited[i]=True
        next_tour = now_tour[:]
        next_tour.append(i)
        next_distance = now_distance + distance(cities[now_place], cities[i])
        tour, min_distance = run_brute (cities, tour, next_tour, next_distance, i, visited, count_visited+1, min_distance)
        visited[i]=False
        
    return tour, min_distance


# In[12]:


# method for creating an initial answer and calling functions to improve it
def solve(cities):
    tour, total_distance = greedy(cities)
    start_time = time.time()
    time_limit = 1
    initial_distance = total_distance
    while (time.time()-start_time<time_limit):
        tour, total_distance = annealing(cities, tour, total_distance, initial_distance, start_time, time_limit)
    tour, total_distance = noCross(cities, tour, total_distance)
    tour, total_distance = brute(cities, tour, total_distance)
    #from IPython.core.debugger import Pdb; Pdb().set_trace()
    return tour


# In[13]:


# reading input from csv file
def readInput(count):
    with open("input_"+str(count)+".csv", encoding='utf-8') as csvfile:
        cities = []
        csvreader = csv.reader(csvfile)
        next(csvreader)  # Skip the header row
        for row in csvreader:
            x, y = float(row[0]), float(row[1])
            cities.append([x, y])
    
    return cities


# In[14]:


# writing the answer as a csv file
def writeOutput(tour, count):
    
    tour = tour[:-1]
    
    with open('output_'+str(count)+".csv", 'w', newline='') as csvfile:
        csvwriter = csv.writer(csvfile)
        # Write the header
        csvwriter.writerow(['index'])
        # Write the tour indices
        for location in tour:
            csvwriter.writerow([location])


# In[15]:


if __name__ == '__main__':
#     for i in range (7):
#         cities = readInput(i)
#         tour = solve(cities)
#         writeOutput(tour, i)
#         print("Finished "+str(i)+"th data!")

# When testing for one particular test case
    i = 5
    cities = readInput(i)
    tour = solve(cities)
    writeOutput(tour, i)
    print("Finished "+str(i)+"th data!")

File: week6/Homework/Homework/Homework.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <ctime>
#include <fstream>
#include <sstream>
#include <algorithm>
#include <iomanip>
using namespace std;

class Solver {
    public:
        double minDistance = 1000000000;
        double initialDistance = 1000000000;
        vector<vector<double>> cities;
        vector<int> answer;
        
        //Start the process of solving
        int start() {
            /*for (int i = 0; i < 8; i++) {
                initialize();
                readInput(i);
                solve();
                writeOutput(i);
                cout << "Finished " << i << "th data!" << endl;
                cout << endl << "-----" << endl << endl;
            }*/

            // When testing for one particular test case
            
            int i = 7;
            initialize();
            readInput(i);
            solve();
            writeOutput(i);
            cout << "Finished " << i << "th data!" << endl;
            

            return 0;
        }

        //Initialize global variables
        void initialize() {
            minDistance = 1000000000;
            initialDistance = 1000000000;
            cities.clear();
            answer.clear();
        }

        //Call each methods to get the minimum distance
        void solve() {

            vector<int> initialTour = greedy();
            //vector<int> initialTour = createInitialTour(); //FOR TESTING

            if (initialTour.size() < 10) {
                initialTour = bruteForce(initialTour);
                answer = initialTour;
            } else {
                int clusterCount = 4;
                initialTour = annealing(initialTour);
                vector<vector<int>> dividedTour = divideTour(initialTour, clusterCount);

                for (int i = 0; i < dividedTour.size(); i++) {
                    dividedTour[i] = removeCross(dividedTour[i]); //TODO: cross-cluster
                    cout << "Finished removing crosses!" << endl;
                }

                for (int i = 0; i < dividedTour.size(); i++) {
                    for (int j = 0; j < dividedTour[i].size(); j++) {
                        answer.push_back(dividedTour[i][j]);
                    }
                }
            }
            return;
        }

        // dividing the tour into smaller clusters
        vector <vector<int>> divideTour(vector<int> initialTour, int clusterCount) {
            vector <vector <int>> dividedTour;
            if (cities.size() > 8000) {
                int length = initialTour.size() / clusterCount;
                length++;
                int index = 0;
                for (int i = 0; i < clusterCount; i++) {
                    vector<int> division;
                    for (int j = 0; j < length; j++) {
                        if (j == length - 1) {
                            if (initialTour.size() % clusterCount <= i) {
                                continue;
                            }
                        }
                        division.push_back(initialTour[index]);
                        index++;
                    }
                    dividedTour.push_back(division);
                }
            }
            else {
                dividedTour.push_back(initialTour);
            }

            return dividedTour;
        }

        // writing the answer as a csv file
        void writeOutput(int count) {
            answer.pop_back();

            ofstream file("output_" + to_string(count) + ".csv");
            file << "index" << endl;
            for (int i : answer) {
                file << i << endl;
            }
        }

        // reading input from csv file
        void readInput(int count) {
           ifstream file("input_" + to_string(count) + ".csv");
            string line;
            getline(file, line);  // Skip the header row
            while (getline(file, line)) {
                stringstream lineStream(line);
                string cell;
                vector<double> city;
                while (getline(lineStream, cell, ',')) {
                    city.push_back(stod(cell));
                }
                cities.push_back(city);
            }
        }

        // FOR TEST: creating a initial tour
        vector <int> createInitialTour() {
            int N = cities.size();
            vector<vector<double>> dist(N, vector<double>(N, 0));
            vector<int> tour;
            for (int i = 0; i < N; i++) {
                for (int j = i; j < N; j++) {
                    dist[i][j] = dist[j][i] = distance(cities[i], cities[j]);
                }
            }

            int currentCity = 0;
            tour = { currentCity };
            double distance = 0;

            for (int i = 1; i < N; i++) {
                int nextCity = i;
                distance += dist[currentCity][nextCity];
                tour.push_back(nextCity);
                currentCity = nextCity;
            }

            distance += dist[currentCity][tour[0]];
            tour.push_back(tour[0]);

            cout << "Finished creating initial tour!" << endl;
            return tour;
        }

        // creating an initial answer based on greedy technique
        vector<int> greedy() {
            int N = cities.size();
            vector<vector<double>> dist(N, vector<double>(N, 0));
            vector<int> tour;
            for (int i = 0; i < N; i++) {
                for (int j = i; j < N; j++) {
                    dist[i][j] = dist[j][i] = distance(cities[i], cities[j]);
                }
            }

            int currentCity = 0;
            vector<bool> unvisitedCities(N, true);
            unvisitedCities[0] = false;
            tour = { currentCity };
            double distance = 0;

            for (int i = 1; i < N; i++) {
                int nextCity = -1;
                double minDist = numeric_limits<double>::infinity();
                for (int j = 0; j < N; j++) {
                    if (unvisitedCities[j] && dist[currentCity][j] < minDist) {
                        nextCity = j;
                        minDist = dist[currentCity][j];
                    }
                }
                distance += minDist;
                unvisitedCities[nextCity] = false;
                tour.push_back(nextCity);
                currentCity = nextCity;
            }

            distance += dist[currentCity][tour[0]];
            tour.push_back(tour[0]);

            cout << "Finished greedy!" << endl;
            return tour;
        }

        // calculate a distance between the two points
        double distance(vector<double> city1, vector<double> city2) {
            return sqrt((city1[0] - city2[0]) * (city1[0] - city2[0]) + (city1[1] - city2[1]) * (city1[1] - city2[1]));
        }

        // simulated annealing
        vector<int> runAnnealing(double startTime, double timeLimit, vector<int> tour, int count) {
            random_device rd;
            mt19937 gen(rd());
            uniform_int_distribution<> dis(1, tour.size() - 2);

            int first = dis(gen);
            int second = dis(gen);

            if (first == second) {
                return tour;
            }

            pair<vector<int>, double> result = swapRoute(first, second, tour);
            vector<int> newTour = result.first;
            double newDistance = result.second;

            double randProb = static_cast<double>(rand()) / RAND_MAX;
            double probability = getProbability(startTime, timeLimit, newDistance);

            // FOR DEBUG:
            if (count % 50 == 0) {
                cout << probability << ", " << newDistance << ", " << minDistance << endl << setprecision(10);
            }
            
            //TODO: look up probability curve
            // Annealing1
            //本当はこっちにしたかったけど上手くいかなかった
            /*if (probability > randProb) {
                tour = newTour;
                minDistance = newDistance;
            }*/

            //Annealing2
            if (newDistance < minDistance) {
                tour = newTour;
                minDistance = newDistance;
            }

            return tour;
        }

        vector<int> annealing(vector<int> tour) {
            double startTime = static_cast<double>(clock()) / CLOCKS_PER_SEC;
            double timeLimit = 600;
            initialDistance = minDistance;
            int count = 0;
            while ((static_cast<double>(clock()) / CLOCKS_PER_SEC - startTime) < timeLimit) {
                tour = runAnnealing(startTime, timeLimit, tour, count);
                count += 1;
            }

            cout << "Finished simulated annealing!" << endl;

            return tour;
        }

        // swap two points inside the route
        pair<vector<int>, double> swapRoute(int first, int second, vector<int> tour) {
            vector<int> newTour(tour.size());
            copy(tour.begin(), tour.end(), newTour.begin());
            double newDistance = minDistance;

            double subtract1 = distance(cities[newTour[first]], cities[newTour[first + 1]]);
            double subtract2 = distance(cities[newTour[second]], cities[newTour[second + 1]]);
            double subtract3 = distance(cities[newTour[first]], cities[newTour[first - 1]]);
            double subtract4 = distance(cities[newTour[second]], cities[newTour[second - 1]]);

            swap(newTour[first], newTour[second]);

            double add1 = distance(cities[newTour[first]], cities[newTour[first + 1]]);
            double add2 = distance(cities[newTour[second]], cities[newTour[second + 1]]);
            double add3 = distance(cities[newTour[first]], cities[newTour[first - 1]]);
            double add4 = distance(cities[newTour[second]], cities[newTour[second - 1]]);

            newDistance -= (subtract1 + subtract2 + subtract3 + subtract4);
            newDistance += (add1 + add2 + add3 + add4);

            return make_pair(newTour, newDistance);
        }

        // calculate a probability for simulated annealing
        double getProbability(double startTime, double timeLimit, double newDistance) {
            if (newDistance < minDistance) {
                return 1.0;
            }
            double startTemp = initialDistance * 0.000015;
            double endTemp = initialDistance * 0.0000001;
            double temp = startTemp + (endTemp - startTemp) * ((double)clock() / CLOCKS_PER_SEC / (startTime + timeLimit));
            double probability = exp((minDistance - newDistance) / temp); //TODO: change function
            //FOR DEBUG:
            //cout << temp << ", " << probability << ", " << (minDistance - newDistance) << ", " << (minDistance-newDistance)/temp << endl;
            return probability;
        }

        // when the two lines intersect each other, reorganize the points so that there is no intersections
        vector<int> removeCross(vector<int> tour) {
            bool continueFlag = true;
            while (continueFlag) {
                continueFlag = false;
                int N = tour.size();
                for (int i = 1; i < N - 2; ++i) {
                    for (int j = i + 2; j < N - 1; ++j) {
                        if (judgeCross(i, j, tour)) {
                            pair<vector<int>, double> result = fixChain(i, j, tour);
                            tour = result.first;
                            double tempDistance = result.second;
                            if (tempDistance < minDistance) {
                                continueFlag = true;
                                minDistance = tempDistance;
                            }
                        }
                    }
                }
            }

            return tour;
        }

        // checking whether two lines (created by p1&p2 and p3&p4) has an intersection
        bool judgeCross(int i, int j, vector<int> tour) {
            vector<double> first = cities[tour[i]];
            vector<double> second = cities[tour[j]];
            vector<double> firstNext = cities[tour[i + 1]];
            vector<double> secondNext = cities[tour[j + 1]];

            // x座標による判定
            if (!maxMinCross(first[0], firstNext[0], second[0], secondNext[0])) {
                return false;
            }

            // y座標による判定
            if (!maxMinCross(first[1], firstNext[1], second[1], secondNext[1])) {
                return false;
            }

            double tc1 = (first[0] - firstNext[0]) * (second[1] - first[1]) + (first[1] - firstNext[1]) * (first[0] - second[0]);
            double tc2 = (first[0] - firstNext[0]) * (secondNext[1] - first[1]) + (first[1] - firstNext[1]) * (first[0] - secondNext[0]);
            double td1 = (second[0] - secondNext[0]) * (first[1] - second[1]) + (second[1] - secondNext[1]) * (second[0] - first[0]);
            double td2 = (second[0] - secondNext[0]) * (firstNext[1] - second[1]) + (second[1] - secondNext[1]) * (second[0] - firstNext[0]);
            return tc1 * tc2 <= 0 && td1 * td2 <= 0;
        }

        // checking whether x or y coordinates of two lines (created by p1&p2 and p3&p4) has an intersection
        bool maxMinCross(double p1, double p2, double p3, double p4) {
            double minab = min(p1, p2);
            double maxab = max(p1, p2);
            double mincd = min(p3, p4);
            double maxcd = max(p3, p4);

            if (minab > maxcd || maxab < mincd) {
                return false;
            }

            return true;
        }

        // changing the order of the tour and re-calculate the distance of the given two points
        pair<vector<int>, double> fixChain(int start, int end, vector<int> tour) {
            vector<int> newTour(tour.size());
            copy(tour.begin(), tour.end(), newTour.begin());
            double newDistance = 0;
            for (int i = 0; i < end - start; ++i) {
                newTour[start + i + 1] = tour[end - i];
            }

            int N = tour.size();

            for (int i = 0; i < N - 1; ++i) {
                newDistance += distance(cities[newTour[i]], cities[newTour[i + 1]]);
            }

            return make_pair(newTour, newDistance);
        }

        // do brute-force method if the number of cities is small
        vector<int> bruteForce(vector<int> tour) {
            vector<bool> visited(tour.size(), false);
            visited[0] = true;
            int countVisited = 1;
            double nowDistance = 0;
            vector<int> nowTour = { 0 };
            tour = runBruteForce(nowTour, nowDistance, 0, visited, countVisited, tour);
            cout << "Finished brute force!" << endl;
            return tour;
        }

        vector<int> runBruteForce(vector<int> nowTour, double nowDistance, int nowPlace, vector<bool> visited, int countVisited, vector<int> tour) {
            if (countVisited == tour.size()-1) {
                nowDistance += distance(cities[nowPlace], cities[0]);
                nowTour.push_back(0);
                if (nowDistance < minDistance) {
                    tour = nowTour;
                    minDistance = nowDistance;
                }
                return tour;
            }

            vector<int> newTour(nowTour.size());
            copy(nowTour.begin(), nowTour.end(), newTour.begin());
            vector<bool> newVisited(visited.size());
            copy(visited.begin(), visited.end(), newVisited.begin());

            for (int i = 0; i < tour.size()-1; ++i) {
                if (visited[i]) continue;

                newVisited[i] = true;
                newTour.push_back(i);
                nowDistance += distance(cities[nowPlace], cities[i]);
                tour = runBruteForce(newTour, nowDistance, i, newVisited, countVisited + 1, tour);
                nowDistance -= distance(cities[nowPlace], cities[i]);
                newTour.pop_back();
                newVisited[i] = false;
            }

            return tour;
        }


};

int main() {
    Solver solver;
    solver.start();

    return 0;
}